### DISCUSSION TITLE
How to Register Dialect Correctly for Spring JPA application.

### DISCUSSION THREAD

**Discussion by Aniket-Singla** [Blazebit/blaze-persistence#1868] [Q&A] [answered]
```
@Bean
    public CriteriaBuilderFactory criteriaBuilderFactory() {
        CriteriaBuilderConfiguration config = Criteria.getDefault();
        config.registerDialect("microsoft", new CustomMSSQLDbmsDialect());
        return config.createCriteriaBuilderFactory(entityManagerFactory);
    }
```

I have followed documentation to register a custom dialect. Problem is that functions are created at the time of Criteria.getDefault() call only and registering custom dialect never helps for microsoft. I have overriden String to use varchar(max) in custom dialect but is not getting picked up in  CAST_STRING.

Any help would be appreciated thanks.

Only option I see currently is to rewrite CriteriaBuilderConfigurationImpl.

**Comment by beikov** [comment, marked as answer]
You could also re-register that one function through a custom `CriteriaBuilderConfigurationContributor`, but apart from that, this is unfortunately the only way to achieve this.
I'm open to making the `DbmsDialect` registration more dynamic though so you can achieve what you desire. Can you please explain why you want to use `varchar(max)` instead? If there is a bug in Blaze-Persistence, we can fix that :)

**Comment by Aniket-Singla** [comment]
There are two issues with SQL Server Dialect

1. CAST_STRING uses varchar as dtype. varchar in sql server is only 30 chars in length. If I am writing a CONCAT + CAST_STRING combination CAST_STRING trims the output to 30.
2. CONCAT method in microsoft Dialect is set to use PlusBasedConcatFunction. SqlServer supports both but with different behaviour. + does not handles null well. I have overriden the function to use ConcatFunction instead .

Thanks for answering. As you said I am overriding functions itself as overriding Dialect does not help.

I believe first point should be handled as bug fix. Second point might depends on end users but I believe CONCAT function should be made default for sql server.

**Reply by beikov** [reply]
The semantics of concatenation with `null` is to produce `null` which the `+` operator fulfills. I don't know how the `concat` function works in SQL Server, but I doubt this is the behavior that we want.
As for 1), I agree that this is a bug. Could you please create a GitHub issue for this and maybe also a PR with a fix? Should be as simple as changing from `"varchar"` to `"varchar(max)"`

**Reply by Aniket-Singla** [reply]
I have opened [#1870](https://github.com/Blazebit/blaze-persistence/issues/1870) as Issue. Will create a PR soon.
